### ISSUE TITLE
1600 duplicated commits in the history

### ISSUE THREAD

**Issue by Hornwitser** [x3dom/x3dom#633]
Thursday September 6th 2012, mlimper merged about 1,640 duplicated commits into the master branch in 9f5b09686ff5c70753ec189fa2a429ac73914eb9, (see [this paste](http://pastebin.com/gxkbjY8e) for a visualization) thereby bumping up the number of commits in the repository at the time from about 2,500 to about 4,150.

Who in their right mind thought this was :ok_hand: to keep!?!

**Comment by mlimper** [comment]
Howdy Hornwitser,

thanks for the pointer! This is of course an interesting and important issue.

May I ask you how you come to the conclusion that the problem originated from the commit you mention? In your visualization, this is the commit that merges both "duplicate branches" together - however, one of its parents is itself a merge commit: https://github.com/x3dom/x3dom/commit/c5c205248bf37c5507702741538fc9fb601b66a2

The duplicates reach back until the beginning of X3DOM development using GitHub, it seems - example:
https://github.com/x3dom/x3dom/commit/b1d5501
https://github.com/x3dom/x3dom/commit/349f606

Following your visualization, this pair of commit is actually the first duplicate. Both branches originate from the same commit, which is https://github.com/x3dom/x3dom/commit/f4ece1c77f2b7bcf133a6e9c0ce9fb44ae2f0cc8

Before using GitHub, X3DOM was developed using an SVN repository. Maybe something went wrong during that step, on some developer machine, and the result somehow got into the master branch?
I'm just wondering how it could happen that the two "duplicate branches" were created at that point.

**Comment by mlimper** [comment]
@yjung Do you remember something going wrong at some point? I can't really recall doing something fancy, but that's already almost four years ago, so I'm not sure. Or do you remember anything originating before 2012, such as the shift from Sourceforge / SVN to GitHub, which could have caused such a problem?

**Comment by Hornwitser** [comment]
Nice catch, it seems like I overlooked that there are two merges that pull in this whole bogus log.  One of which is 9f5b096 and the other is as you spotted c5c2052.

Both of these commits are the cause of this problem because they reference the bogus history as one of their parents.  Before these commits were introduced, the bogus history was not an ancestor of the master branch, and thus it didn't exist in the master branch.

I also noticed while scrolling through the history that while they are very similar, they are not identical.  The bogus history is missing quite a few empty commits, as well as the following: 23399bc, 8a34309, 218ace3, ef80200, 9f1e5a8, e02388a.  All relating to the implementation of the "SQUEEZE algorithm", which is then removed for no given reason.

All of this information put together suggests that an attempt was made at rewriting the history of the repository using `git filter-branch`, in order to remove all traces of the SQUEEZE thingy, without entirely understanding the consequences of the actions done.  `git filter-branch` is one of the very few tools git has that can create this kind of duplicated history with changes.
